datarobot_provider/operators/predictions.py

```diff
@@ -0,0 +1,202 @@
+        return predict_job.id
+
+
+class GetPredictionsOperator(BaseDatarobotOperator):
+    """
+    Gets predictions from a PredictJob.
+
+    Args:
+        project_id (str): DataRobot project ID.
+        predict_job_id (str): PredictJob ID.
+
+    Returns:
+        DataFrame: Predictions DataFrame.
+
+    Raises:
+        AirflowFailException: If predict_job_id or project_id is not provided.
+    """
+
+    template_fields: List[str] = ["project_id", "predict_job_id"]
+
+    def __init__(self, *, project_id: str, predict_job_id: str, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.project_id = project_id
+        self.predict_job_id = predict_job_id
+
+    def validate(self) -> None:
+        if not self.predict_job_id or not self.project_id:
+            raise AirflowFailException("predict_job_id and project_id must be provided")
+
+    def execute(self, context: Context) -> DataFrame:
+        predictions = dr.PredictJob.get_predictions(
+            project_id=self.project_id,
+            predict_job_id=self.predict_job_id,
+        )
+
+        self.log.info(f"Predictions retrieved, predict_job_id={self.predict_job_id}")
+
+        ## NOTE: Can we even return a DataFrame here?
+        return predictions
+
+
+class GetPredictJobOperator(BaseDatarobotOperator):
+    """
+    Gets a PredictJob.
+
+    Args:
+        project_id (str): DataRobot project ID.
+        predict_job_id (str): PredictJob ID.
+
+    Returns:
+        dr.PredictJob: PredictJob object.
+
+    Raises:
+        AirflowFailException: If predict_job_id or project_id is not provided.
+        AirflowFailException: If PredictJob has already finished.
+    """
+
+    template_fields: List[str] = ["project_id", "predict_job_id"]
+
+    def __init__(self, *, project_id: str, predict_job_id: str, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.project_id = project_id
+        self.predict_job_id = predict_job_id
+
+    def validate(self) -> None:
+        if not self.predict_job_id or not self.project_id:
+            raise AirflowFailException("predict_job_id and project_id must be provided")
+
+    def execute(self, context: Context) -> dr.PredictJob:
+        try:
+            predict_job = dr.PredictJob.get(
+                project_id=self.project_id,
+                predict_job_id=self.predict_job_id,
+            )
+        except dr.errors.PendingJobFinished:
+            # NOTE: Is this the correct exception to raise in this case?
+            raise AirflowFailException(
+                "PredictJob has already finished, please use PredictJobGetPredictionsOperator to retrieve predictions"
+            ) from dr.errors.PendingJobFinished
+
+        self.log.info(f"PredictJob retrieved, predict_job_id={self.predict_job_id}")
+
+        # NOTE: Can we return the PredictJob object here?
+        return predict_job
+
+
+class AwaitAndGetPredictionsOperator(BaseDatarobotOperator):
```

Comment:
Unfortunately, I think this operator should not exist, since it's against the Airflow design. We should have a [Sensor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/sensors.html) which just waits for completion instead. So the flow would be this:

`RequestPredictionsOperator -> WaitForPredictionsSensor -> GetPredictionsOperator`

The `GetPredictionsOperator` should store predictions somewhere, I'm not sure where. If we want to save in Catalog, then we should have a `SavePredictionsToCatalogOperator` or smth like that.

Reply:
Hmm, yeah I think that makes more sense. We can use the `PredictJob.wait_for_completion` in the `WaitForPredictionsSensor` sensor.

For the "predictions storage" question, I think we have 2 options depending on our use cases:

1. Have a standard place to store the predictions ALWAYS and return some sort of id of that - that makes it easier for users to not worry about managing another resource just for their predictions
2. Write separate operators to store the predictions in separate places like s3, catalog, or some other kind of store -> I don't think that's a good idea as it will get quite complicated very quickly - so unless there's a compelling use case, we should avoid doing this

Reply:
I'm gonna go ahead and create that sensor and disassemble this operator.